master/internal/resourcemanagers/priority.go

```diff
@@ -0,0 +1,331 @@
+)
+
+type priorityScheduler struct {
+	preemptionEnabled bool
+}
+
+// NewPriorityScheduler creates a new scheduler that schedules tasks via priority.
+func NewPriorityScheduler(config *SchedulerConfig) Scheduler {
+	return &priorityScheduler{preemptionEnabled: config.Priority.Preemption}
+}
+
+func (p *priorityScheduler) Schedule(rp *ResourcePool) ([]*AllocateRequest, []*actor.Ref) {
+	return p.prioritySchedule(rp.taskList, rp.groups, rp.agents, rp.fittingMethod)
+}
+
+func (p *priorityScheduler) prioritySchedule(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+) ([]*AllocateRequest, []*actor.Ref) {
+	agentsSplitByLabel := splitAgentsByLabel(agents)
+	toAllocate := make([]*AllocateRequest, 0)
+	toRelease := make([]*actor.Ref, 0)
+
+	// Since labels are a hard scheduling constraint, process every
+	// label independently.
+	for label, agentsWithLabel := range agentsSplitByLabel {
+		toAllocatedForLabel, toReleaseForLabel := p.priorityScheduleByLabel(
+			taskList, groups, agentsWithLabel, fittingMethod, label)
+		toAllocate = append(toAllocate, toAllocatedForLabel...)
+		toRelease = append(toRelease, toReleaseForLabel...)
+	}
+
+	return toAllocate, toRelease
+}
+
+func (p *priorityScheduler) priorityScheduleByLabel(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+	label string,
+) ([]*AllocateRequest, []*actor.Ref) {
+	// All pending zero slot tasks get scheduled right away.
+	toAllocate := getAllPendingZeroSlotTasks(taskList, label)
+	toRelease := make(map[*actor.Ref]bool)
+
+	// Sort tasks by priorities and timestamps. This sort determines the order in which
+	// tasks are scheduled and preempted.
+	priorityToPendingTasksMap, priorityToScheduledTaskMap := sortTasksByPriorityAndTimestamp(
+		taskList, groups, label)
+
+	// Make a local copy of the agent state that we will modify.
+	localAgentsState := deepCopyAgents(agents)
+
+	// Once we are unable to start a task of a higher priority, do not start anymore tasks.
+	startTasks := true
+
+	for _, priority := range getOrderedPriorities(priorityToPendingTasksMap) {
+		allocationRequests := priorityToPendingTasksMap[priority]
+		log.Debugf("processing priority %d with %d pending tasks",
+			priority, len(allocationRequests))
+
+		successfulAllocations, unSuccessfulAllocations := trySchedulingPendingTasksInPriority(
+			allocationRequests, localAgentsState, fittingMethod)
+
+		// Only add these tasks to the lists of tasks to start if all tasks of higher priority
+		// have been scheduled.
+		if startTasks {
+			for _, allocatedTask := range successfulAllocations {
+				log.Infof("scheduled task: %s", allocatedTask.Name)
+			}
+			toAllocate = append(toAllocate, successfulAllocations...)
+		}
+
+		// All pending tasks in this priority were successfully scheduled.
+		if len(unSuccessfulAllocations) == 0 {
+			continue
+		}
+		startTasks = false
```

Comment:
q: it looks like `startTasks` won't be changed back to `true` even if all unsuccessful allocations are placed on preempted tasks. Is it the expected behavior?